File: app/src/main/java/de/herrmann_engel/rbv/utils/SwipeEvents.java
package de.herrmann_engel.rbv.utils;

import android.view.MotionEvent;
import android.view.View;

public class SwipeEvents implements View.OnTouchListener {

    private static final int SWIPE_DISTANCE_MIN = 150;
    private static final int SWIPE_START_DISTANCE_MIN = 100;
    private static final int MOVE_DIRECTION_RIGHT = 1;
    private static final int MOVE_DIRECTION_LEFT = 2;
    private static final int MOVE_DIRECTION_BOTTOM = 3;
    private static final int MOVE_DIRECTION_TOP = 4;

    private float initialY;
    private float initialX;
    private int moveDirection;

    public SwipeEvents() {
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        int action = event.getActionMasked();
        if (action == MotionEvent.ACTION_DOWN) {
            initialX = event.getX();
            initialY = event.getY();
            moveDirection = 0;
            return true;
        } else if (action == MotionEvent.ACTION_MOVE) {
            float distanceX = event.getX() - initialX;
            float distanceY = event.getY() - initialY;
            if (Math.abs(distanceX) > Math.abs(distanceY)) {
                if (Math.abs(distanceX) < SWIPE_DISTANCE_MIN) {
                    if (moveDirection == 0) {
                        if (distanceX > 0 && initialX < SWIPE_START_DISTANCE_MIN) {
                            moveDirection = MOVE_DIRECTION_RIGHT;
                        } else if (initialX > v.getWidth() - SWIPE_START_DISTANCE_MIN) {
                            moveDirection = MOVE_DIRECTION_LEFT;
                        }
                        return true;
                    } else if (distanceX > 0 && moveDirection == MOVE_DIRECTION_RIGHT || distanceX < 0 && moveDirection == MOVE_DIRECTION_LEFT) {
                        onMoveX(distanceX);
                        return true;
                    }
                } else if (distanceX > 0 && moveDirection == MOVE_DIRECTION_RIGHT) {
                    moveDirection = -1;
                    onSwipeRight();
                    return true;
                } else if (distanceX < 0 && moveDirection == MOVE_DIRECTION_LEFT) {
                    moveDirection = -1;
                    onSwipeLeft();
                    return true;
                }
            } else {
                if (Math.abs(distanceY) < SWIPE_DISTANCE_MIN) {
                    if (moveDirection == 0) {
                        if (distanceY > 0 && initialY < SWIPE_START_DISTANCE_MIN) {
                            moveDirection = MOVE_DIRECTION_BOTTOM;
                        } else if (initialY > v.getHeight() - SWIPE_START_DISTANCE_MIN) {
                            moveDirection = MOVE_DIRECTION_TOP;
                        }
                        return true;
                    } else if (distanceY > 0 && moveDirection == MOVE_DIRECTION_BOTTOM || distanceY < 0 && moveDirection == MOVE_DIRECTION_TOP) {
                        onMoveY(distanceY);
                        return true;
                    }
                } else if (distanceY > 0 && moveDirection == MOVE_DIRECTION_BOTTOM) {
                    moveDirection = -1;
                    onSwipeBottom();
                    return true;
                } else if (distanceY < 0 && moveDirection == MOVE_DIRECTION_TOP) {
                    moveDirection = -1;
                    onSwipeTop();
                    return true;
                }
            }
        }
        onMoveCancel();
        return false;
    }

    public void onMoveCancel() {
        moveDirection = -1;
    }

    public void onMoveX(float distance) {
    }

    public void onMoveY(float distance) {
    }

    public void onSwipeRight() {
    }

    public void onSwipeLeft() {
    }

    public void onSwipeTop() {
    }

    public void onSwipeBottom() {
    }

}

File: app/src/main/java/de/herrmann_engel/rbv/utils/SwipeEvents.kt
package de.herrmann_engel.rbv.utils

import android.annotation.SuppressLint
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import kotlin.math.abs

open class SwipeEvents : OnTouchListener {
    private var initialY = 0f
    private var initialX = 0f
    private var moveDirection = 0

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouch(v: View, event: MotionEvent): Boolean {
        val action = event.actionMasked
        if (action == MotionEvent.ACTION_DOWN) {
            initialX = event.x
            initialY = event.y
            moveDirection = 0
            return true
        } else if (action == MotionEvent.ACTION_MOVE) {
            val distanceX = event.x - initialX
            val distanceY = event.y - initialY
            if (abs(distanceX) > abs(distanceY)) {
                if (abs(distanceX) < SWIPE_DISTANCE_MIN) {
                    if (moveDirection == 0) {
                        if (distanceX > 0 && initialX < SWIPE_START_DISTANCE_MAX) {
                            moveDirection = MOVE_DIRECTION_RIGHT
                        } else if (distanceX < 0 && initialX > v.width - SWIPE_START_DISTANCE_MAX) {
                            moveDirection = MOVE_DIRECTION_LEFT
                        }
                        return true
                    } else if (distanceX > 0 && moveDirection == MOVE_DIRECTION_RIGHT || distanceX < 0 && moveDirection == MOVE_DIRECTION_LEFT) {
                        onMoveX(distanceX)
                        return true
                    }
                } else if (distanceX > 0 && moveDirection == MOVE_DIRECTION_RIGHT) {
                    moveDirection = -1
                    onSwipeRight()
                    return true
                } else if (distanceX < 0 && moveDirection == MOVE_DIRECTION_LEFT) {
                    moveDirection = -1
                    onSwipeLeft()
                    return true
                }
            } else {
                if (abs(distanceY) < SWIPE_DISTANCE_MIN) {
                    if (moveDirection == 0) {
                        if (distanceY > 0 && initialY < SWIPE_START_DISTANCE_MAX) {
                            moveDirection = MOVE_DIRECTION_BOTTOM
                        } else if (distanceY < 0 && initialY > v.height - SWIPE_START_DISTANCE_MAX) {
                            moveDirection = MOVE_DIRECTION_TOP
                        }
                        return true
                    } else if (distanceY > 0 && moveDirection == MOVE_DIRECTION_BOTTOM || distanceY < 0 && moveDirection == MOVE_DIRECTION_TOP) {
                        onMoveY(distanceY)
                        return true
                    }
                } else if (distanceY > 0 && moveDirection == MOVE_DIRECTION_BOTTOM) {
                    moveDirection = -1
                    onSwipeBottom()
                    return true
                } else if (distanceY < 0 && moveDirection == MOVE_DIRECTION_TOP) {
                    moveDirection = -1
                    onSwipeTop()
                    return true
                }
            }
        }
        onMoveCancel()
        return false
    }

    open fun onMoveCancel() {
        moveDirection = -1
    }

    open fun onMoveX(distance: Float) {}
    open fun onMoveY(distance: Float) {}
    open fun onSwipeRight() {}
    open fun onSwipeLeft() {}
    open fun onSwipeTop() {}
    open fun onSwipeBottom() {}

    companion object {
        private const val SWIPE_DISTANCE_MIN = 150
        private const val SWIPE_START_DISTANCE_MAX = 50
        private const val MOVE_DIRECTION_RIGHT = 1
        private const val MOVE_DIRECTION_LEFT = 2
        private const val MOVE_DIRECTION_BOTTOM = 3
        private const val MOVE_DIRECTION_TOP = 4
    }
}
